### PR TITLE
Support Vimeo elements in AMP

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

Adds support for Vimeo embeds in AMP 🎥 ⚡️ 

Vimeo and YouTube embeds are constructed in a very similar way so I have built one consolidated cleaner for both embed types. 

## What is the value of this and can you measure success?

Articles with Vimeo embeds will now be indexable by AMP search platforms. Users will be able to view articles containing Vimeo embeds through AMP, providing a faster mobile experience for these articles.

## Does this affect other platforms - Amp, Apps, etc?

Yep 😄 

## Screenshots

picture 122

## Request for comment

@guardian/dotcom-platform

### CODE REVIEW

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 1]
```diff
@@ -39,25 +39,47 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
+      video.attr("layout", "responsive")
+    }
+
+    def getVideoIdFromUrl(videoUrl: String): Option[String] = {
+      val youtubePattern = ".*youtube.com/watch\\?v=(.+).*".r
+      val vimeoPattern = ".*vimeo.com/(\\d+).*".r
+      videoUrl match {
+        case youtubePattern(videoId) => Some(videoId)
+        case vimeoPattern(videoId) => Some(videoId)
+        case _ => None
```
Comment (jfsoul) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: Can we combine this with the `getElementTypeByUrl` regex?  I think we should be able to do one regex and return something that encapsulates both the element name and the video ID.

File: tools/amp-validation/endpoints.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ module.exports = (() => {
         '/sport/2016/aug/25/katie-ledecky-first-pitch-washington-nationals-bryce-harper', // Instagram + Twitter
         '/sport/live/2016/aug/25/county-cricket-surrey-lancashire-live-blog', // Twitter
         //'/sport/blog/2016/aug/25/baseball-olympic-games-2020-tokyo-mlb-team-selection', // Vine
-        //'/stage/2016/aug/24/they-drink-it-in-the-congo-review-adam-brace', // Vimeo
+        '/stage/2016/aug/24/they-drink-it-in-the-congo-review-adam-brace', // Vimeo
```
Comment (jfsoul) [inline tools/amp-validation/endpoints.js]: 👏

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 3]
```diff
@@ -16,6 +16,61 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
...
+    val youtubeUrl = "https://www.youtube.com/watch?v=foo_12-34"
+    val doc = s"""<html><body><figure class="element-video" data-canonical-url="$youtubeUrl"><iframe></iframe></figure></body></html>"""
+    val document: Document = Jsoup.parse(doc)
+    val result: Document = clean(document)
+
+    result.getElementsByTag("amp-youtube").size should be(1)
+  }
+
+  "AmpEmbedCleaner" should "not create an amp-youtube element if videoid missing" in {
+    val youtubeUrl = s"http://www.youtube.com/"
```
Comment (jfsoul) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: Looks like we have `http` and `https` for youtube in these tests, which makes me think: maybe we should only be accepting `https` video links?

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 4]
```diff
@@ -16,6 +16,61 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
...
+  "AmpEmbedCleaner" should "be able to create an amp-youtube element and an amp-vimeo in the same document" in {
+    val youtubeUrl = "https://www.youtube.com/watch?v=foo_12-34"
+    val vimeoUrl = "https://vimeo.com/1234"
+    val doc = s"""<html><body><figure class="element-video" data-canonical-url="$youtubeUrl"><iframe></iframe></figure><figure class="element-video" data-canonical-url="$vimeoUrl"><iframe></iframe></figure></body></html>"""
+    val document: Document = Jsoup.parse(doc)
+    val result: Document = clean(document)
+
+    result.getElementsByTag("amp-youtube").size should be(1)
+    result.getElementsByTag("amp-vimeo").size should be(1)
+  }
```
Comment (jfsoul) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: everybody loves tests! 💯 

could add some tests for edge cases with the regex if you felt it was required (might not be).  E.g. `youtubercom.com`, `youtube.com.ru`, `fakeyoutube.com` etc.

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala (hunk 1 above)
Comment (TBonnin) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: Are they referring to the same element, if not I will keep them separate?

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala (hunk 3 above)
Comment (SiAdcock) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: It doesn't look like it matters. Even if the canonical URL is `http`, `amp-youtube` and `amp-vimeo` still request the `https` version of the video.

picture 123

Comment (SiAdcock) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: I'll create specific `http` and `https` tests since either protocol can be entered Composer.

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala (hunk 1 above)
Comment (SiAdcock) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: They are testing the same element. I'll try to combine them and return an instance of a class that contains the `videoId` and the `elementType`

**Comment by SiAdcock** [comment]
Thanks for your suggestions, @jfsoul and @TBonnin 😄  Would you care to take one more look?

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 5]
```diff
@@ -39,14 +39,16 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
 
+  class AmpExternalVideo(val videoId: String, val elementType: String) {}
+
   object AmpExternalVideo {
-    def getElementTypeByUrl(videoUrl: String) : String = {
-      val youtubePattern = ".*youtube.com.*".r
-      val vimeoPattern = ".*vimeo.com.*".r
+    def getAmpExternalVideoByUrl(videoUrl: String) : Option[AmpExternalVideo] = {
+      val youtubePattern = ".*https?://www.youtube.com/watch\\?v=(.+).*".r
+      val vimeoPattern = ".*https?://vimeo.com/(\\d+).*".r
```
Comment (jfsoul) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: Not sure this regex is quite right

- Probably start with `^` instead of `.*`
- I think there are a few `/` and `.` that need escaping
- I think the query param ? should be `\?`, it looks like it's escaped twice
- Can we capture the youtube video id with `.*v=([\d\w]+)&?.*` - otherwise we might capture other query params before or after video id.

I might be off mark on some of this as I don't have good knowledge of youtube/vimeo urls and their edge cases.  Let me know what you think.

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 6]
```diff
@@ -39,14 +39,16 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
     })
   }
 
+  class AmpExternalVideo(val videoId: String, val elementType: String) {}
```
Comment (jfsoul) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: case class AmpExternalVideo(val videoId: String, val elementType: String)

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 7]
```diff
@@ -39,14 +39,16 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
+    def getAmpExternalVideoByUrl(videoUrl: String) : Option[AmpExternalVideo] = {
+      val youtubePattern = ".*https?://www.youtube.com/watch\\?v=(.+).*".r
+      val vimeoPattern = ".*https?://vimeo.com/(\\d+).*".r
       videoUrl match {
-        case youtubePattern() => "amp-youtube"
-        case vimeoPattern() => "amp-vimeo"
-        case _ => ""
+        case youtubePattern(videoId) => Some(new AmpExternalVideo(videoId, "amp-youtube"))
+        case vimeoPattern(videoId) => Some(new AmpExternalVideo(videoId, "amp-vimeo"))
+        case _ => None
```
Comment (jfsoul) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: With the case class you can avoid the `new` here.

You could go further and introduce `case class YoutubeExternalVideo(videoId: String) extends AmpExternalVideo("amp-youtube")` or something which would require some other changes... but I think your simple solution is better in this case.

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 8]
```diff
@@ -16,6 +16,79 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
     document
   }
 
+  private def documentWithVideos(videoUrls: Seq[String]): Document = {
+    val doc = "<html><body>" +
+      videoUrls.foldLeft(""){ (res: String, next: String) => res + s"""<figure class="element-video" data-canonical-url="$next"><iframe></iframe></figure>""" } +
```
Comment (TBonnin) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: Rather than using `foldLeft` you can simple `map` your `Seq` of urls into a `Seq` of html strings and then use `mkString` to join them

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 9]
```diff
@@ -16,6 +16,79 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
     document
   }
 
+  private def documentWithVideos(videoUrls: Seq[String]): Document = {
```
Comment (TBonnin) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: Nitpicking: you can use `varargs` so you don't have to pass a `Seq` but multiple `Strings`
ie:
```
private def documentWithVideos(videoUrls: String*): Document
....
documentWithVideos("https://vimeo.com/")
documentWithVideos("https://www.youtube.com/watch?v=foo_12-34","https://vimeo.com/1234")

```

**Review by TBonnin — CHANGES_REQUESTED** [review]

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala (hunk 5 above)
Comment (johnduffell) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: use triple double quotes to stop scala compiler looking inside strings, then you don't have to escape for the compiler's benefit
```scala
"""my \?lovely "string" of joy"""
```

Comment (SiAdcock) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: Cool, that's a top tip! Thanks @johnduffell and @jfsoul 👍 💎 ❤️

**Review by jfsoul — APPROVED** [review]
Great work! Thanks for the changes 💯

**Review by TBonnin — APPROVED** [review]

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 10]
```diff
@@ -39,25 +39,40 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
-        youtube.attr("height", "3")
-        youtube.attr("layout", "responsive")
-        element.appendChild(youtube)
+  sealed abstract class AmpExternalVideo(val videoId: String, val elementType: String)
+  case class YoutubeExternalVideo(override val videoId: String) extends AmpExternalVideo(videoId, "amp-youtube")
+  case class VimeoExternalVideo(override val videoId: String) extends AmpExternalVideo(videoId, "amp-vimeo")
+
+  object AmpExternalVideo {
+    def getAmpExternalVideoByUrl(videoUrl: String) : Option[AmpExternalVideo] = {
+      val youtubePattern = """^https?:\/\/www\.youtube\.com\/watch\?v=([^#\^&\^?]+).*""".r
```
Comment (johnduffell) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: parser combinators are the answer! https://github.com/sirthias/parboiled2
